ewatercycle/models/abstract.py

```diff
@@ -16,6 +16,9 @@ class AbstractModel(metaclass=ABCMeta):
     def __init__(self):
         self.bmi: Bmi = None  # bmi should set in setup() before calling its methods
 
+    def __del__(self):
+        self.bmi: Bmi = None
```

Comment:
I amnot sure why you assign the `self.bmi` to `None` rather than deleting it like `del self.bmi`.  In the case of deleting the` self.bmi`, you can use `pytest.raises(AttributeError)` in  `test_finalize_resets_bmi` as:

```python
    with pytest.raises(AttributeError):
        model.bmi
```

Reply:
Maybe here we could as well use `del self.bmi`, but testing this doesn't work with `pytest.raises`, since if the `model` instance is deleted, you have no reference to the `bmi` object anymore.

Reply:
By setting it to None, bmi object will have a reference count of zero and be cleaned up by Pythons garbage collector at a later point of time. While calliing `del self.bmi` will call the container termination code immediately.

Reply:
Are you sure about that @sverhoeven ? The [docs on `__del__`](https://docs.python.org/3/reference/datamodel.html?highlight=del#object.__del__) state that

> `del x` doesn’t directly call `x.__del__()` — the former decrements the reference count for `x` by one, and the latter is only called when `x`’s reference count reaches zero.

I'm also fine with using `del` instead of setting it to None, though. Just thought it was nice to keep the attribute even if there's no bmi object currently assosciated with it.